### PR TITLE
Update Travis script to fail if leaks detected

### PR DESCRIPTION
**Pull request by davidbeckingsale**

### CODE REVIEW

File: tests/unit/resource/resource_tests.cpp [hunk 1]
```diff
@@ -42,6 +42,8 @@ TEST(DefaultMemoryResource, AllocateDeallocate)
   auto alloc = std::make_shared<umpire::resource::DefaultMemoryResource<TestAllocator> >(umpire::Platform::cpu, "TEST", 0, umpire::MemoryResourceTraits{});
```
Comment (jdahm) [inline tests/unit/resource/resource_tests.cpp]: Totally orthogonal to this, but we don't need the shared pointer here.

Comment (davidbeckingsale) [reply tests/unit/resource/resource_tests.cpp]: 💯

File: scripts/travis/build_and_test.sh [hunk 2]
```diff
@@ -37,7 +37,13 @@ if [[ "$DO_BUILD" == "yes" ]] ; then
       or_die ctest -T test --output-on-failure -V
     fi
     if [[ "${DO_MEMCHECK}" == "yes" ]] ; then
-      or_die ctest -T memcheck
+      regex="^Memory Leak - [1-9]+"
+      while read -r line; do
+          if [[ "$line" =~ $regex ]]; then
+            echo "Found leaks: $line"
```
Comment (mcfadden8) [inline scripts/travis/build_and_test.sh]: Not really sure this is sufficient.  [https://travis-ci.com/LLNL/Umpire/jobs/207604652](https://travis-ci.com/LLNL/Umpire/jobs/207604652) reports a leak, but I cannot tell which leak it is?  Could we say what the line number is so we can go back in the build log and look for context?

Comment (davidbeckingsale) [reply scripts/travis/build_and_test.sh]: The problem is that you can't get context without dumping all the valgrind logs. I could have it do that on error?

Comment (mcfadden8) [reply scripts/travis/build_and_test.sh]: I think that would be good.  I am not sure what _Memory Leak - 8_ means.

Comment (mcfadden8) [reply scripts/travis/build_and_test.sh]: Since it is reporting one now, can we try it and see what it looks like?

Comment (davidbeckingsale) [reply scripts/travis/build_and_test.sh]: It means there are 8 leaks 🙄

Comment (mcfadden8) [reply scripts/travis/build_and_test.sh]: Well that was obvious.  Would love to see what they are :-)

Comment (mcfadden8) [reply scripts/travis/build_and_test.sh]: Much better!

File: src/umpire/ResourceManager.cpp [hunk 3]
```diff
@@ -163,52 +169,47 @@ ResourceManager::initialize()
...
-  m_allocators_by_id[host_allocator->getId()] = host_allocator;
-
-  m_default_allocator = host_allocator;
-
-#if defined(UMPIRE_ENABLE_DEVICE)
-  auto device_allocator = m_memory_resources[resource::Device];
-  m_allocators_by_name["DEVICE"] = device_allocator;
-  m_allocators_by_id[device_allocator->getId()] = device_allocator;
+  {
+    auto allocator{registry.makeMemoryResource("DEVICE", getNextId())};
```
Comment (jdahm) [inline src/umpire/ResourceManager.cpp]: This is going to cause problems with GCC 4.9.3.

File: CMakeLists.txt [hunk 4]
```diff
@@ -89,7 +89,7 @@ if (NOT BLT_LOADED)
   include(${BLT_SOURCE_DIR}/SetupBLT.cmake)
 endif()
 
-include(cmake/SetupCompilerFlags.cmake)
```
Comment (mcfadden8) [inline CMakeLists.txt]: Where did the pedantic warnings go?

**Comment by jdahm** [comment]
Some performance numbers from LC on this branch:
```Malloc/malloc/4                      78 ns         78 ns    8983648
Malloc/malloc/8                      78 ns         78 ns    8982935
Malloc/malloc/64                     77 ns         77 ns    9067000
Malloc/malloc/512                    95 ns         95 ns    7344900
Malloc/malloc/1024                  154 ns        154 ns    4544862
Malloc/free/4                        54 ns         54 ns   12868088
Malloc/free/8                        54 ns         54 ns   12868020
Malloc/free/64                       54 ns         54 ns   12873621
Malloc/free/512                      63 ns         63 ns   11186809
Malloc/free/1024                     74 ns         74 ns    9518914
FixedMallocPool/allocate/4            9 ns          9 ns   81367016
FixedMallocPool/deallocate/4          8 ns          8 ns   92529088
Host/allocate/4                     288 ns        288 ns    2379057
Host/allocate/8                     288 ns        288 ns    2430870
Host/allocate/64                    284 ns        284 ns    2463373
Host/allocate/512                   321 ns        321 ns    2187596
Host/allocate/1024                  376 ns        376 ns    1864352
Host/deallocate/4                   313 ns        313 ns    2237299
Host/deallocate/8                   313 ns        313 ns    2239217
Host/deallocate/64                  265 ns        265 ns    2653869
Host/deallocate/512                 266 ns        266 ns    2629635
Host/deallocate/1024                281 ns        281 ns    2493700
PoolHost/allocate/4                 454 ns        454 ns    1531950
PoolHost/allocate/8                 454 ns        454 ns    1523855
PoolHost/allocate/64                490 ns        490 ns    1476274
PoolHost/allocate/512               461 ns        461 ns    1509496
PoolHost/allocate/1024              507 ns        507 ns    1343538
PoolHost/deallocate/4              1051 ns       1050 ns     670326
PoolHost/deallocate/8               928 ns        928 ns     743320
PoolHost/deallocate/64              734 ns        734 ns     981486
PoolHost/deallocate/512             623 ns        623 ns    1007250
PoolHost/deallocate/1024            620 ns        620 ns     997630
FixedPoolHost/allocate/4            495 ns        495 ns    1418292
FixedPoolHost/deallocate/4          201 ns        201 ns    3498888
```
I have been waiting to no avail for 10 minutes to get PoolHost benchmarks to finish on develop.

**Pull request merged**